src/services/swap_service.ts

```diff
@@ -315,6 +326,63 @@ export class SwapService implements ISwapService {
             swapQuoteRequestOpts = ASSET_SWAPPER_MARKET_ORDERS_OPTS;
         }
 
+        const marketSide = sellAmount !== undefined ? MarketOperation.Sell : MarketOperation.Buy;
+        let feeToken = buyToken; // Default fee token is buy token
+
+        // Prepare Sell Token Fees
+        let sellTokenFees: Fees | undefined = undefined; // fees object to return to the caller
+        let sellTokenFeeAmount = ZERO; // total sell token fee to charge
+        let sellTokenFeeAmounts: AffiliateFeeAmount[] = []; // sell token fee amounts used by affiliate fee transformer
+        let sellTokenFeeOnChainTransferGas = ZERO; // the gas cost for transferring sell token fee on-chain
+        if (metaTransactionVersion !== undefined && marketSide === MarketOperation.Sell) {
+            // Use sell token as fee
+
+            // Narrow the type
+            if (!sellAmount) {
+                throw new Error('sellAmount is undefined when market direction is sell');
+            }
+
+            feeToken = sellToken;
+            if (!providedGasPrice) {
+                providedGasPrice = await this._swapQuoter.getGasPriceEstimationOrThrowAsync();
+            }
+
+            // Only get `sellTokenAmountPerWei` when necessary
+            const sellTokenAmountPerWei = feeConfigs?.gasFee
+                ? await this._swapQuoter.getTokenAmountPerWei(sellToken, {})
+                : undefined;
+            // Need to calculate all fees in order to adjust `sellAmount` before passing it down to sampler & router
+            ({
+                sellTokenFees,
+                sellTokenFeeAmount,
+                sellTokenFeeOnChainTransfers: sellTokenFeeAmounts,
+                sellTokenFeeOnChainTransferGas,
+            } = this._getSellTokenFees(
+                feeConfigs,
+                sellToken,
+                sellAmount,
+                sellTokenAmountPerWei,
+                providedGasPrice,
+                AVG_MULTIPLEX_TRANFORM_ERC_20_GAS, // use historic average gas cost for multiplex & transformERC20
+                metaTransactionVersion === 'v1' ? TRANSFER_GAS : TRANSFER_FROM_GAS, // meta-transaction v1 uses affiliate fee transformer which calls `transformerTransfer` for on-chain fee transfer.
+                // meta-transaction v2 uses `FixinTokenSpender._transferERC20TokensFrom`.
+            ));
+
+            if (sellAmount.lte(sellTokenFeeAmount)) {
```

Comment:
under what condition can this happen? (is this something that should never happen?)

Reply:
This can happen if the fees taken are greater than the `sellAmount` specified by the user. The most likely scenario would be if the trade size (say $1) is really small on mainnet, then gas fee cost (part of `sellTokenFeeAmount`) is very likely to be greater than `sellAmount`. I will add a comment on this condition